File: 215-kth-largest-element-in-an-array/215-kth-largest-element-in-an-array.cpp
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        //implementing a min heap so that we can only keep k elements in the heap
        priority_queue<int, vector<int>, greater<int>> pq;
        for(int n : nums)
        {
            pq.push(n);
            if(pq.size() > k)  //O(log k) time 
            {
                pq.pop();  //remove the min element in the pq at that time
            }
        }
        return pq.top();
    }
};

File: 215-kth-largest-element-in-an-array/215-kth-largest-element-in-an-array.py
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        k = len(nums) - k
        
        def quickselect(l, r):
            p, pivot = l, nums[r]
            for i in range(l, r):
                if(nums[i] <= pivot):
                    nums[i], nums[p] = nums[p], nums[i]
                    p += 1
            nums[p], nums[r] = nums[r], nums[p]
            
            if(k < p):
                return quickselect(l, p-1)
            elif(k > p):
                return quickselect(p+1, r)
            else:
                return nums[p]
        
        return quickselect(0, len(nums)-1)